lib/src/code/code.dart

```diff
@@ -2,55 +2,84 @@ import 'package:collection/collection.dart';
 import 'package:flutter/material.dart';
 import 'package:highlight/highlight_core.dart';
 
+import '../named_sections/named_section.dart';
+import '../named_sections/parsers/abstract.dart';
 import '../single_line_comments/parser/single_line_comment_parser.dart';
 import '../single_line_comments/parser/single_line_comments.dart';
+import '../single_line_comments/single_line_comment.dart';
 import 'code_line.dart';
 import 'text_range.dart';
 import 'tokens.dart';
 
 class Code {
   final String text;
   final List<CodeLine> lines;
+  final Map<String, NamedSection> namedSections;
 
   factory Code({
     required String text,
     Result? highlighted,
     Mode? language,
+    AbstractNamedSectionParser? namedSectionParser,
+    Set<String> readOnlySectionNames = const {},
   }) {
     final sequences = SingleLineComments.byMode[language] ?? [];
+
+    final commentParser = SingleLineCommentParser.parseHighlighted(
+      text: text,
+      highlighted: highlighted,
+      singleLineCommentSequences: sequences,
+    );
+
     final lines = _textToCodeLines(
       text: text,
       highlighted: highlighted,
       language: language,
-      singleLineCommentSequences: sequences,
+      commentsByLines: commentParser.getCommentsByLines(),
+    );
+
+    final sections = namedSectionParser?.parse(
+          singleLineComments: commentParser.comments,
+          lineCount: lines.length,
+        ) ??
+        const [];
+    final sectionsMap = {for (final s in sections) s.name: s};
+
+    _applyNamedSectionsToLines(
```

Comment:
can this be a pure function with no side effects?

Reply:
This will run on each change in the editor, so I don't want to slow it down.

For the same reason I also think of making a `CodeLine` mutable to just change its `isReadOnly` in there, but then we can't override `hashCode` and `operator ==`, so I have not yet decided on this.

Anyway tests are there to compensate for the risks of side effects.